nixos/pppd: init

danderson opened the pull request:
###### Motivation for this change

pppd is used to control a lot of home internet connections. I want to use NixOS as my home router, and I need pppd to bring up my FTTH internet connection. It will also be useful to people using DSL (PPPoE or PPPoA) and classic dial-up internet.

pppd is already available in nix as the `ppp` derivation. It's used by other nixos modules like `pptpd`, but there's no module for running PPP sessions directly. This module fixes that.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

(Other tests)
- [x] Configured and ran a pppd service, and successfully got internet access using this module.

###### Notify maintainers

cc @falsifian (because you're the maintainer of the `ppp` package, which this module service-ifies)

danderson commented:
One style question I didn't know how to answer: what should the commit message be? CONTRIBUTING.md says that it should be `init at <version>`, but this commit only adds a nixos module, which doesn't have a version.

alexarice commented:
The commit title is fine

alexarice commented on maintainers/maintainer-list.nix:
```diff
@@ -1439,6 +1439,12 @@
     githubId = 245394;
     name = "Hannu Hartikainen";
   };
+  danderson = {
```
The addition to `maintainer-list.nix` should be in a separate commit titled `maintainers: add danderson`

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
```
It is usual to put 
```
let
  cfg = config.services.pppd;
in
```
Here and then use that below

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+    services.pppd = {
+      package = mkOption {
+        default = pkgs.ppp;
+        defaultText = "pkgs.ppp";
```
Is this necessary, I thought the document system could cope with this

danderson replied:
I don't know, I was copying other nixos modules that do this with their pkg reference. I'm happy to remove it and see what happens when the docs go live?

alexarice replied:
It can't hurt so leave it

alexarice replied:
There is probably a way to locally build the docs but I'm not quite sure how if you wanted to check

danderson replied:
I'll look around tomorrow for a way to build the options search engine locally and check for this.

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+
+      peers = mkOption {
+        default = {};
+        type = types.loaOf (types.submodule (
```
`loaOf` should be being deprecated. See https://github.com/NixOS/nixpkgs/pull/63103 . I would use `attrsOf` here but if you might need multiple submodules per name (or if ordering matters) then use list

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+          {
+            options = {
+              name = mkOption {
+                example = "dialup";
```
Think you need `default = name;` here or am I mistaken?

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+                description = "Name of the PPP peer.";
+              };
+
+              enable = mkOption {
```
You could use `mkEnableOption` here

alexarice replied:
Ah you want it to be on by default, this is probably fine

danderson replied:
Yes, in this case I think the default should be true because it only matters if the submodule exists. So the most likely use case for this option is "I want to temporarily disable this 1 peer".

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+                description = "Whether the PPP session is automatically started at boot time.";
+              };
+
+              config = mkOption {
```
Could consider using a structured config. See https://github.com/NixOS/rfcs/pull/42

danderson replied:
pppd configurations are a bit unusual compared to more modern configs, so I'm not sure if it fits cleanly inside a structured config. For example, here's my pppd config:

```
plugin rp-pppoe.so eno1
name "redacted-isp-username"
noipdefault
persist
defaultroute
hide-password
noauth
```

I'm not sure how I would convert this to a structured config, given the irregular shape. Do you have a suggestion?

alexarice replied:
I'm not sure, was more just something to think about

danderson replied:
I'll think about it some more tomorrow. In the meantime, I think it's easy to switch from a text config to structured later and deprecate this field.

danderson replied:
After reviewing the RFC, I believe there is no obvious mapping between pppd configs and nix. The format is just too weird and irregular.

However, per the RFC's recommendations for "unsuitable formats", I changed the type of `config` to `types.lines`, so at least specifying additional configuration works reasonably.

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+            };
+
+            config = {
+              name = mkDefault name;
```
I would prefer `default = name` as I have written above I think

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+    etcFiles = map mkEtc enabledConfigs;
+    systemdConfigs = map mkSystemd enabledConfigs;
+
+  in {
```
```suggestion
  in mkIf cfg.enable {
```

I think you should include a global enable option

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,137 @@
+  in {
+    environment.etc = mkMerge etcFiles;
+    systemd.services = mkMerge systemdConfigs;
+  };
```
Should the package be added to the environment or is that not necessary (i.e if it is only run by systemd)

danderson replied:
I don't know what the custom is for NixOS. The only reason to invoke `pppd` directly is debugging a connection, but you can do that easily with `systemctl restart pppd-peer` as well. Based on that, I thought I would not pollute the system $PATH with a "useless" daemon, but it's not a strong opinion. WDYT?

alexarice replied:
The only reason I would consider it is that adding `pppd` to `environment.systemPackages` won't add your wrapped version (which I'm guessing is needed to debug). It is probably fine not to have it

danderson replied:
Good point!

IMO, it's still easy to debug with `systemctl restart pppd-name`, so having `pppd` or the wrapper in `systemPackages` is just more noise for nothing. And it's easier to add it later if we're wrong, than to remove it and potentially break someone who depends on it. So I propose we start without adding to `systemPackages`, and see how that goes.

danderson committed:
maintainers: add danderson

danderson committed:
nixos/pppd: init

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,141 @@
+
+    enabledConfigs = filter (f: f.enable) (attrValues cfg.peers);
+
+    mkEtc = cfg: {
```
May be worth renaming this variable

alexarice commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,141 @@
+          RuntimeDirectory = "pppd";
+          RuntimeDirectoryPreserve = true;
+        };
+        wantedBy = optional peerCfg.autostart "multi-user.target";
```
I think I would prefer `wantedBy = mkIf peerCfg.autostart [ "multi-user.target" ]` but your call

alexarice reviewed (changes requested):
Sorry I didn't notice this earlier. You haven't added you're module to `module-list.nix` so it won't be available.

danderson commented:
Doh! I've been testing on my router with an import directly in configuration.nix, that's why. Added to module-list.nix. Thanks for catching that!

flokli commented:
@danderson can you also add a small nixos test with 2 vms ping each other through a ppp(d) connection?

danderson commented:
Sure! I've never set up a pppoe+pppd server, so it might take me a bit, but I'm working on it.

danderson committed:
nixos/tests/pppd: init

This test creates a PPPoE link between two machines, and verifies
that the machines can ping each other.

danderson commented:
I've added a test as requested. One VM runs a PPPoE server, and the other uses this pppd module to connect as a client. Test verifies bidirectional ping reachability through the PPP link. PTAL?

flokli commented on nixos/tests/pppd.nix:
```diff
@@ -0,0 +1,62 @@
+import ./make-test.nix (
+  let
+    chap-secrets = {
+      text = ''"flynn" * "reindeerflotilla" *'';
```
I assume this is not a real password :-)

danderson replied:
You assume correctly :) It's Flynn's password from [Tron](https://en.wikipedia.org/wiki/Tron). It's a classic "I need a quick password for testing" :P

flokli commented:
@GrahamcOfBorg build nixosTests.pppd

flokli commented on nixos/modules/services/networking/pppd.nix:
```diff
@@ -0,0 +1,141 @@
+        wants = [ "network.target" ];
+        after = [ "network-pre.target" ];
+        serviceConfig = {
+          ExecStart = "${getBin pppd-wrapped}/sbin/pppd call ${peerCfg.name} nodetach nolog";
```
`pppd-wrapped` can be removed, by setting `"${peerCfg.name}".environment` - we don't use pppd-wrapped outside the systemd unit anyway.

danderson replied:
Good call! Initially I wasn't sure if I wanted pppd in systemPackages, but we decided we only care about systemd units, so this is much cleaner.

Done. Nixos test still passes, and my home internet still works.

flokli commented:
@GrahamcOfBorg test pppd

danderson commented:
Anything else we need before merging this? The PR history says I requested a review from @Infinisil . I don't _think_ I did? But I can't edit the reviewer list, and I welcome more eyes on this if that's the next step :)

flokli commented:
@danderson Infinisil is marked as a codeowner for the module system.

I wanted to await the nixos test results from CI, and forgot about it.

Thanks for the ping!
